ipa-core/src/protocol/ipa_prf/oprf_padding/mod.rs

```diff
@@ -1,5 +1,814 @@
+{
+    let initial_len = input.len();
+
+    // H1 and H2 add padding noise
+    input = apply_dp_padding_pass::<C, BK, TV, TS, B>(
+        ctx.narrow(&PaddingDpStep::PaddingDpPass1),
+        input,
+        Role::H1,
+        Role::H2,
+        Role::H3,
+        &padding_params,
+    )
+    .await?;
+
+    // H3 and H1 add padding noise
+    input = apply_dp_padding_pass::<C, BK, TV, TS, B>(
+        ctx.narrow(&PaddingDpStep::PaddingDpPass2),
+        input,
+        Role::H3,
+        Role::H1,
+        Role::H2,
+        &padding_params,
+    )
+    .await?;
+
+    // H2 and H3 add padding noise
+    input = apply_dp_padding_pass::<C, BK, TV, TS, B>(
+        ctx.narrow(&PaddingDpStep::PaddingDpPass3),
+        input,
+        Role::H2,
+        Role::H3,
+        Role::H1,
+        &padding_params,
+    )
+    .await?;
+
+    let after_padding_len = input.len();
+    tracing::info!(
+        "Total number of padding records added: {}. Padding Parameters: {:?}",
+        after_padding_len - initial_len,
+        padding_params
+    );
+
+    Ok(input)
+}
+
+/// Apply dp padding with one pair of helpers generating the noise
+/// Steps
+///     1.  Helpers `h_i` and `h_i_plus_one` will get the same rng from PRSS
+///         and use it to sample the same random noise for padding from `OPRFPaddingDp`.
+///         They will generate secret shares of these fake rows.
+///     2.  `h_i` and `h_i_plus_one` will send the send `total_number_of_fake_rows` to `h_out`
+///     3.  `h_out` will generate secret shares of zero for as many rows as the `total_number_of_fake_rows`
+///
+/// # Errors
+/// Will propogate errors from `OPRFPaddingDp`
+/// # Panics
+/// Will panic if called with Roles which are not all unique
+pub async fn apply_dp_padding_pass<C, BK, TV, TS, const B: usize>(
+    ctx: C,
+    mut input: Vec<OPRFIPAInputRow<BK, TV, TS>>,
+    h_i: Role,
+    h_i_plus_one: Role,
+    h_out: Role,
+    padding_params: &PaddingParameters,
+) -> Result<Vec<OPRFIPAInputRow<BK, TV, TS>>, Error>
+where
+    C: Context,
+    BK: BooleanArray + U128Conversions,
+    TV: BooleanArray,
+    TS: BooleanArray,
+{
+    // assert roles are all unique
+    assert!(h_i != h_i_plus_one);
+    assert!(h_i != h_out);
+    assert!(h_out != h_i_plus_one);
+
+    let mut total_number_of_fake_rows = 0;
+    let mut padding_input_rows: Vec<OPRFIPAInputRow<BK, TV, TS>> = Vec::new();
+
+    // Step 1: Helpers `h_i` and `h_i_plus_one` will get the same rng from PRSS
+    // and use it to sample the same random noise for padding from OPRFPaddingDp.
+    // They will generate secret shares of these fake rows.
+    if ctx.role() != h_out {
+        total_number_of_fake_rows = two_parties_add_dummies::<C, BK, TV, TS, B>(
+            &ctx,
+            &mut padding_input_rows,
+            h_i,
+            h_i_plus_one,
+            padding_params,
+        )?;
+    }
+
+    // Step 2: h_i and h_i_plus_one will send the send total_number_of_fake_rows to h_out
```

Comment:
This constitutes *some* kind of leakage, even if it's not concerning. It would be cool to have some kind of a comment either explaining why this is OK, or pointing to some resource explaining why this is OK.

Reply:
Okay, will point to the `oprf_padding/README.md` and make sure it is more fully explained there.